### ISSUE TITLE
Client camel cases headers

### ISSUE THREAD

**Issue by baritonehands** [http-kit/http-kit#286]
When sending an http request using the client, it automatically camel cases all the headers. However, that's not the behavior I want in my app. My server is expecting lower case headers.

From HttpClient.exec:
```HeaderMap headers = HeaderMap.camelCase(cfg.headers);```

There needs to be either a way to turn off the feature per request, or camel case only the sensitive headers that some servers can't handle.

**Comment by ptaoussanis** [comment]
Is there a reason you wouldn't want your server to accept headers with any casing?

**Comment by ptaoussanis** [comment]
As an aside - and I'm kind of picking on you randomly here (long week, sorry) - please consider whether it might be possible to rephrase issues like this in a slightly more courteous way in future? For example: with a hi, a thank you, and with your concern in the form of a question instead of "there needs to be...".

Big deal? No. Just been noticing a lot of this lately, and know more than a few people going through various stages of open source fatigue.

As I mentioned above, http headers are case insensitive + it's generally standard procedure (and good hygiene) for servers to normalise casing for incoming headers that they care about.

Does that make sense / seem reasonable?

Cheers :-)

**Comment by baritonehands** [comment]
Apologies for curtness, I have opinions about whether we should continue with the standard Hi <name>, ... Sincerely <name> form in this internet age, but that's a discussion for another day. However, I definitely could have asked instead of demanded, and was a little lazy to do so, so again I apologize.

First of all, this use case is an experimental project for me, which may never see the light of day. Take that as an indication of priority.

I think my issue boils down to the fact that camel casing the headers was unexpected. I entered them in lowercase, and was surprised to find the undocumented (unless I missed it) behavior that they all were converted to be camel cased. I had to read the source code to determine if it was my code or not.

The HTTP spec does say that header names are case insensitive, but the implementation is not consistent across many common servers. Passing "host" instead of "Host" for example causes issues with some server implementations. I think it's actually not an easy problem because there's not necessarily a simple drop-in case-insensitive hashmap implementation in a particular server's language. 

I ended up using clj-http instead. But, just this moment I looked at the clj-http code and discovered that it lower-cases all headers. It happens to work in my case, but it also is not what I think is the ideal behavior.

I think the preferred implementation is to pass along the developer's intended case, or at least provide the option to turn off the auto camel-case. But it's also extremely likely that changing that behavior would break someone else's code when it stops camel casing.

I leave it up to you what you think is best. For what it's worth, I think the performance numbers for http-kit are impressive, and it's well designed. I will use it for other projects, but this was a blocker for me for this particular project because of the sheer amount of case sensitive code I might have to change.

**Comment by baritonehands** [comment]
Sorry, did it again.

Thanks for considering this change. :)

**Comment by venantius** [comment]
Hm. 

On the one hand, I think I agree with the expectation that a given HTTP client (ours) not do any unexpected transformations on data passed into a given function. At the very least, it indicates we're currently doing un-necessary work in the client. 

On the other hand, since you can't really rely on any given client's behavior w/r/t the casing of headers per the HTTP spec, I'm not sure you should be engineering your server to expect a specific casing. 

In other words, I agree with you in the general case but not for your specific case :).

**Comment by ptaoussanis** [comment]
> Thanks for considering this change. :)

No problem, appreciate the clarification :-)

> The HTTP spec does say that header names are case insensitive, but the implementation is not consistent across many common servers. 

That sort of goes to the definition of the spec saying that headers are case insensitive though. Servers cannot and should not try to rely on consistent casing. If a server cares about a header, by the spec, it must normalize that header.

Afaik every real server does this, including Ring by default.

As for how the header keys are normalized, that varies - some will do all upper, some all lower, some camel case.

> Passing "host" instead of "Host" for example causes issues with some server implementations.

Not sure about the specific example you have in mind, but that'd seem to be in violation of the http spec.

To clarify: sometimes a library API will promise to normalize the headers _for you_, so that you can treat them all as having a specific case for convenience.

>  I think it's actually not an easy problem because there's not necessarily a simple drop-in case-insensitive hashmap implementation in a particular server's language.

Sorry, again not sure how that's a problem if you normalize casing. Make every incoming header lowercase, and you can use your hashmap.

> But, just this moment I looked at the clj-http code and discovered that it lower-cases all headers.

That'll usu. be standard behaviour. As you've seen, the particular normalization will vary though.

> I think the preferred implementation is to pass along the developer's intended case, or at least provide the option to turn off the auto camel-case.

The danger is there that people might be inclined to try and rely on the casing, which is something we should discourage.

> For what it's worth, I think the performance numbers for http-kit are impressive, and it's well designed. I will use it for other projects, but this was a blocker for me for this particular project because of the sheer amount of case sensitive code I might have to change.

Thank you. Just to clarify: I didn't author http-kit's, and it's not my project. Just try help maintain it a little as time allows (along with many other folks).

Re: this request- our options would be to change the default or make this configurable. Both of those would be a misstep, I believe. Really would suggest that following best practices here makes sense for good reasons; just normalize incoming header keys on your server side.

That way, your server won't depend on any particular (unspecified) behaviour, and will be compatible with the rest of the internet.

Does that seem reasonable?

**Comment by ptaoussanis** [comment]
Hi @venantius 

> At the very least, it indicates we're currently doing un-necessary work in the client.

Possibly, but don't think that that would be a significant amount of work by any measure - and there might be reasons why casing _needs_ to be normalized with the current implementation. Would need to check, but really suspect it wouldn't be worth anyone's time.

**Comment by baritonehands** [comment]
I did find some examples for my points about inconsistent implementations, but I think many of those were old posts and have since been fixed.

I think it's reasonable not to change. We should be handling these as case insensitive.

I did have more what ifs that were interesting to me, but probably not common enough:
1. What if a server is not controlled by us, expects headers in a specific case and we need to work around it?
2. Even if I did control the code, what if I couldn't convince my organization that it should be changed?
3. What if the capitalization of these headers carries some human-friendly semantic meaning? E.g. DRY-Header vs. Dry-Header. In one case, capitalization implies an acronym, but the software can handle either. It's only for the developer to understand it better. (A real stretch, but other libraries don't handle this case either).

**Comment by ptaoussanis** [comment]
> I did have more what ifs that were interesting to me, but probably not common enough:

Appreciate the examples. In all of these cases I would suggest: we have an unambiguous spec, which does clearly state that header keys must be case insensitive. Any server not conforming, is unambiguously in violation of the spec and therefore buggy.

Should we try write code that can interoperate with other code that's buggy? In principle yes, probably, if the cost isn't too high.

Here we get into subjective territory, but I'd argue that the cost in this case may be high if it ends up encouraging fragile design that's in contravention of a ~universally accepted spec.

Just a heads-up that I'll personally be signing off from this issue, have my hands pretty full with some urgent work atm. Please feel free to continue the discussion w/o me if anything still needs to be resolved.

Best of luck, cheers! :-)
